### PR TITLE
WIP: Cldc 571 allow users to login and change password

### PR DESCRIPTION
**Pull request by dushan-madetech**

### CODE REVIEW

File: config/initializers/devise.rb [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+  # ==> Controller configuration
+  # Configure the parent class to the devise controllers.
+  # config.parent_controller = 'DeviseController'
+
+  # ==> Mailer Configuration
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = "please-change-me-at-config-initializers-devise@example.com"
```
Comment (baarkerlounger) [inline config/initializers/devise.rb]: Do we know who this should be coming from?

File: spec/support/controller_macros.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+module ControllerMacros
+  # def login_admin
+  #   before(:each) do
+  #     @request.env["devise.mapping"] = Devise.mappings[:admin]
+  #     sign_in FactoryBot.create(:admin) # Using factory bot as an example
+  #   end
+  # end
+
+  def login_user
```
Comment (baarkerlounger) [inline spec/support/controller_macros.rb:9]: This is very nice. Would have expected the namespace to be more like `/spec/helpers/login_helper` or something like that though maybe?

File: config/initializers/devise.rb (hunk 1 above)
Comment (dushan-madetech) [reply config/initializers/devise.rb]: No but I can find out. I assume it's the email address we want registration and password related emails to be sent from so I suspect that Sona or Sarah might have an answer or will be able to point me in the right direction

File: spec/support/controller_macros.rb (hunk 2 above)
Comment (dushan-madetech) [reply spec/support/controller_macros.rb:9]: I was following the advice in this article https://yuta-san.medium.com/a-simple-login-test-with-rspec-devise-factorybot-in-rails-29aeb2ebc4ab which I think is based on the information contained in here https://github.com/heartcombo/devise/wiki/How-To:-Test-controllers-with-Rails-%28and-RSpec%29

Comment (dushan-madetech) [reply spec/support/controller_macros.rb:9]: I'm happy to change this if you think it should be changed 😄

File: lib/tasks/form_definition.rake [hunk 3]
```diff
@@ -1,50 +1,51 @@
 require "json"
 require "json-schema"
-
+# rubocop:disable Lint/ShadowingOuterLocalVariable
```
Comment (baarkerlounger) [inline lib/tasks/form_definition.rake]: Would suggest reverting this, it's been resolved in main now

File: spec/rails_helper.rb [hunk 4]
```diff
@@ -113,4 +113,10 @@
   # Silence capybara logging puma start up messages to stdout on first js: true
   # spec
   Capybara.server = :puma, { Silent: true }
+
+  # For Devise > 4.1.1
+  config.include Devise::Test::ControllerHelpers, type: :controller
+  config.include Devise::Test::IntegrationHelpers, type: :request
+  # Use the following instead if you are on Devise <= 4.1.1
```
Comment (baarkerlounger) [inline spec/rails_helper.rb]: Can probably remove these comments? Not ever going to be relevant to us I think

File: spec/support/devise.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+require_relative "./controller_macros"
+
+RSpec.configure do |config|
+  # For Devise > 4.1.1
```
Comment (baarkerlounger) [inline spec/support/devise.rb]: Same for these

**Review by baarkerlounger — APPROVED** [review]

**Review by baarkerlounger — APPROVED** [review]

**Pull request merged**